crates/cext/src/dag.rs

```diff
@@ -0,0 +1,189 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+
+use qiskit_circuit::bit::{ClassicalRegister, QuantumRegister};
+use qiskit_circuit::dag_circuit::DAGCircuit;
+
+/// @ingroup QkDag
+/// Construct a new empty DAG.
+///
+/// You must free the returned DAG with qk_dag_free when done with it.
+///
+/// @return A pointer to the created DAG.
+///
+/// # Example
+///
+///     QkDag *empty = qk_dag_new();
+///
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_dag_new() -> *mut DAGCircuit {
+    let dag = DAGCircuit::new();
+    Box::into_raw(Box::new(dag))
+}
+
+/// @ingroup QkDag
+/// Add a quantum register to the DAG.
+///
+/// @param dag A pointer to the DAG.
+/// @param reg A pointer to the quantum register.
+///
+/// # Example
+/// ```c
+///     QkDag *dag = qk_dag_new();
+///     QkQuantumRegister *qr = qk_quantum_register_new(1024, "my_register");
+///     qk_dag_add_quantum_register(dag, qr);
+///     qk_quantum_register_free(qr);
+///     qk_dag_free(dag);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag`` and
+/// if ``reg`` is not a valid, non-null pointer to a ``QkQuantumRegister``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_add_quantum_register(
+    dag: *mut DAGCircuit,
+    reg: *const QuantumRegister,
+) {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let dag = unsafe { mut_ptr_as_ref(dag) };
+    let qreg = unsafe { const_ptr_as_ref(reg) };
+
+    dag.add_qreg(qreg.clone())
+        .expect("Invalid register unable to be added to DAG");
+}
```

Comment:
This is consistent with `QkCircuit` so we shouldn't change it, though imo it's a shame we didn't do something like
```rust
pub unsafe extern "C" fn qk_dag_add_quantum_register(
    dag: *mut DAGCircuit,
    reg: *mut QuantumRegister,
) {
    let qr = MaybeUninit::uninit();
    unsafe { ptr::swap(qr.as_mut_ptr(), reg) };
    unsafe { mut_ptr_as_ref(dag).add_qreg(qr.assume_init()) }
        .expect("caller is responsible for passing non-conflicting registers");
}
```
and take ownership of the `QuantumRegister` so that C doesn't have to manually free it.

It doesn't really make much difference, since registers are just `Arc`s internally, but this arguably did want to be move semantics from C.

Reply:
I saw that, but I didn't comment because what you wrote is still valid - you can drop them in either order.